Implement network healthcheck

p-strusiewiczsurmacki-mobica opened the pull request:
This PR implements networking healthcheck as described in #31 and #32.
It also fixes #41.

CPI taint will be removed once all networking checks pass. This is done after first reconciliation loop iteration for 'full' pods and at the end of main() function for bpf-only pods.

Implemented checks are:

- FRR - checks if FRR daemon is active and in running state (will not be used on bpf-only pods)
- Interfaces - checks if provided interfaces are in UP state
- Reachability - checks if TCP connection can be established to provided hosts on selected ports (note: if connection is refused, but host is reachable then test will still pass). Connection timeout can be configured.

All data can be provided using _optional_ config file. Config file location is by default `/opt/network-operator/net-healthcheck-config.yaml` but can be changed with env variable `OPERATOR_NETHEALTHCHECK_CONFIG`. However changing path will make the config file mandatory.

Additionally unit tests were created for the new `healthcheck` package.

schrej commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -0,0 +1,355 @@
+	var _ = Context("LoadConfig() should", func() {
+		var _ = It("return error if config is mandatory but does not exist", func() {
+			t := testing.T{}
+			t.Setenv(CONFIG_ENV, "/some/invalid/path")
```
I think constructing your own `testing.T` doesn't work. Just use `Expect(os.Setenv()).To(Succeed())` here instead

p-strusiewiczsurmacki-mobica replied:
Done

schrej commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -0,0 +1,355 @@
+		var _ = It("return no error if valid config exists", func() {
+			filePath, err := os.CreateTemp(tmpPath, "healtchecktest")
+			Expect(err).ToNot(HaveOccurred())
+			_, err = filePath.WriteString(fakeConf)
```
for the fakeconf you could also just create a `testdata` directory in this package, put it in there and commit to git, and then load using a relative path

p-strusiewiczsurmacki-mobica replied:
Done

Did not want to add irrelevant files to the repository, but if you say it's OK, then it's simpler approach, which I like.

schrej commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -0,0 +1,355 @@
+	})
+})
+
+type fakeFRRManagerError struct{}
```
instead of writing mocks yourself, you could also have a look at https://github.com/uber-go/mock

It's pretty nice, and allows to register handlers to return custom values for testing and so on, so you only need to generate one mock and can use it for everything. Might be a bit overkill here though.

p-strusiewiczsurmacki-mobica replied:
OK, I've removed my mocks and uber-go/mock is now used.

schrej replied:
Would be good to add a `// go:generate <gomock command>` somewhere, maybe in proximity of those interfaces. Then we can just run `go generate ./...` to regenerate if necessary

p-strusiewiczsurmacki-mobica replied:
Done

schrej commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -0,0 +1,355 @@
+		})
+		var _ = It("return error when trying to remove taint (update node)", func() {
+			c := fake.NewSimpleClientset(fakeNodes)
+			c.CoreV1().(*fakecorev1.FakeCoreV1).PrependReactor("update", "nodes",
```
controller-runtime also has a pretty nice fakeclient if you are able to use that one

p-strusiewiczsurmacki-mobica replied:
Done

Although, had to add a mock for this client's error, as can't find anything similar to `PrependReactor` available in goclient.

schrej commented on pkg/reconciler/reconciler.go:
```diff
@@ -57,6 +62,29 @@ func NewReconciler(client client.Client, anycastTracker *anycast.AnycastTracker)
+		return nil, err
+	}
+
+	c, err := clientgo.NewForConfig(cfg)
```
same as above, consider using the controller-runtime client

p-strusiewiczsurmacki-mobica replied:
Done.

Used reconciler.client for this one as it's already created.

p-strusiewiczsurmacki-mobica commented on main.go:
```diff
@@ -159,6 +162,47 @@ func main() {
+			os.Exit(1)
+		}
+
+		c, err := clientgo.NewForConfig(cfg)
```
Yes, 'veI tried to use it, but it seems that controller-runtime's client cannot be used before mgr.Start() call.
When I try to use it in main function (for bpf-only node) it returns
`{"error": "the cache is not started, can not read objects"}` when trying to get the node's object from the API.

p-strusiewiczsurmacki-mobica replied:
Done

p-strusiewiczsurmacki-mobica commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -0,0 +1,211 @@
+		client:        client,
+		ctx:           ctx,
+		isInitialized: false,
+		Logger:        log.FromContext(ctx),
```
Changed that to `log.Log.WithName("HealthCheck")`

p-strusiewiczsurmacki-mobica commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -0,0 +1,211 @@
+	}
+	if !isMandatory && err != nil {
+		return config, nil
+	}
```
Reworked that. Added additional bool variable and made the if easier to understand, I think.

p-strusiewiczsurmacki-mobica commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -0,0 +1,211 @@
+func (hc *HealthChecker) CheckReachability() error {
+	for _, i := range hc.netConfig.Reachability {
+		if err := hc.checkReachabilityItem(i); err != nil {
+			if strings.Contains(err.Error(), "refused") {
```
Done

p-strusiewiczsurmacki-mobica commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -0,0 +1,355 @@
+}
+
+var _ = Describe("HealthCheck", func() {
+	var _ = Context("LoadConfig() should", func() {
```
I believe that I've seen linter that was not happy about unhandled return values before. Anyway, changed that as I can't see similar issue with current golangci-lint setup.

schrej commented on main.go:
```diff
@@ -159,6 +162,41 @@ func main() {
+	if onlyBPFMode {
+		clusterClient, err := client.New(clientConfig, client.Options{})
+		if err != nil {
+			setupLog.Error(err, "problem kreating controller-runtime client")
```
```suggestion
			setupLog.Error(err, "problem creating controller-runtime client")
```

schrej commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -0,0 +1,240 @@
+	return NewHealthCheckToolkit(frr, netlink.LinkByName, tcpDialer)
+}
+
+//go:generate mockgen -source=../../pkg/healthcheck/healthcheck.go -destination=../../pkg/healthcheck/mock/mock_healthcheck.go
```
Can you try the following here?
```
//go:generate mockgen -destination ../mock/mock_healtheck.go . frrInterface,tcpDialerInterface
```
alternatively this could be in something like a `doc.go` file in the `mock` package so the command is closer to the generated code. In that case the `.` would need to be replaced by the full package path (`github.com/telekom/das-schiff-network-operator/pkg/healthcheck`)

p-strusiewiczsurmacki-mobica replied:
Done. :) Had to export the interfaces though.

schrej reviewed (approved):
Two more nits, the rest looks good from my side, thank you!

I'll leave it to Christopher for a second review.

chdxD1 reviewed (commented):
LGTM so far. I would like to have some kind of retry, especially for the connectivity step.

Also what I forgot: We need to come up with a way to specify interfaces names per node because they might differ (usually on masters we have eno1, eno2, on workers the PCIe cards might be on different slots, depending on HW type, resulting in different ens<slot> names)

p-strusiewiczsurmacki-mobica commented:
@chdxD1 
> LGTM so far. I would like to have some kind of retry, especially for the connectivity step.

Just added retries for the reachability test. Updated readme so you can quickly check if it's what you were thinking about. Additionally fixed a bug I found in the reachability test.

Should I add some sleep between retries?

> Also what I forgot: We need to come up with a way to specify interfaces names per node because they might differ (usually on masters we have eno1, eno2, on workers the PCIe cards might be on different slots, depending on HW type, resulting in different ens names)

As for the interfaces names - it is currently configured per node as the config file is separate for each node. But it's not the most user-friendly way I think. Maybe some configmaps would be better? If so, maybe it could be done in another PR?

chdxD1 commented:
@p-strusiewiczsurmacki-mobica missed that config per node thing, I initially assumed that it was part of the overall config. It is probably fine for now, other solutions would be a bit more complicated. We can provide the config as part of our deployment process!

watzkuh commented on README.md:
```diff
@@ -87,6 +87,39 @@ An internal loop is tracking these interfaces and reapplies the eBPF code when n
+There are 3 types of checks implemented currently:
+ - FRR check - will check if FRR's daemon is up and running. This check will be only used in pods that DO NOT work in bpf-only mode.
+ - interfaces - will check if all network interfaces that are expected to are up
+ - reachability - will check if TCP connection to provided hosts can be achieved. Each host will be queired `retries` times before reporting failure.
```
```suggestion
 - reachability - will check if TCP connection to provided hosts can be achieved. Each host will be queried `retries` times before reporting failure.
```

watzkuh commented on README.md:
```diff
@@ -87,6 +87,39 @@ An internal loop is tracking these interfaces and reapplies the eBPF code when n
+ - interfaces - will check if all network interfaces that are expected to are up
+ - reachability - will check if TCP connection to provided hosts can be achieved. Each host will be queired `retries` times before reporting failure.
+
+ Interfaces and reachability checks can be configured using config file located in `/opt/network-operator/net-healthcheck-config.yaml` on the node.
```
```suggestion
 Interfaces and reachability checks can be configured through a config file located in `/opt/network-operator/net-healthcheck-config.yaml` on the node.
```

watzkuh commented on README.md:
````diff
@@ -87,6 +87,39 @@ An internal loop is tracking these interfaces and reapplies the eBPF code when n
+ The syntax of the config file is as follow:
+
+ ```
+interfaces: // list of interfaces that should have UP state
+  - eth0
+reachability: // list of hosts and ports to query in reachability check
+- host: "host.domain.com"
+  port: 80
+- host: "192.168.1.100"
+  port: 8400
+timeout: 5s // timeout for reachability test (if not provided, default 3s will be used)
+retries: 10 // idicates how many times each host provided in reachibility list will be queried before failure will be reported, default 3
````
```suggestion
interfaces: # list of interfaces that should have UP state
  - eth0
reachability: # list of hosts and ports to query in reachability check
- host: "host.domain.com"
  port: 80
- host: "192.168.1.100"
  port: 8400
timeout: 5s # timeout for reachability test (if not provided, default 3s will be used)
retries: 10 # indicates how many times each host provided in reachibility list will be queried before failure will be reported, default 3
```

watzkuh commented on README.md:
````diff
@@ -87,6 +87,39 @@ An internal loop is tracking these interfaces and reapplies the eBPF code when n
+retries: 10 // idicates how many times each host provided in reachibility list will be queried before failure will be reported, default 3
+ ```
+
+Location of the networking healthcheck config file can be altered by setting `OPERATOR_NETHEALTHCHECK_CONFIG` enviromental vairable on the node to selected path, e.q.
````
```suggestion
The location of the networking healthcheck config file can be altered by setting the `OPERATOR_NETHEALTHCHECK_CONFIG` environment variable on the node to the desired path, e.g.:
```

watzkuh commented on README.md:
````diff
@@ -87,6 +87,39 @@ An internal loop is tracking these interfaces and reapplies the eBPF code when n
+OPERATOR_NETHEALTHCHECK_CONFIG=/etc/network-operator/my-healtcheck-config.yaml
+```
+
+> NOTE: Please be aware that custom config file location has to be mounted into the network-operator pod.
````
```suggestion
> NOTE: Please be aware that the custom config file location has to be mounted into the network-operator pod.
```

watzkuh commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -0,0 +1,269 @@
+		if err != nil {
+			timeout = time.Second * defaultTCPTimeout
+		} else {
+			timeout = time.Second * time.Duration(seconds)
```
How hard to we want to try to "silently do the right thing" here?
Might be worth to at least log a warning that we can't deal with the config the user provided (and will now just assume they probably meant seconds / fallback to the default timeout).

p-strusiewiczsurmacki-mobica replied:
> How hard to we want to try to "silently do the right thing" here? 

I'm completely open to suggestions on that. Maybe others would like to give their insights?

Anyway, added logging for now to give user more info on what's going on. That's really good suggestion, btw!

watzkuh commented on pkg/reconciler/reconciler.go:
```diff
@@ -86,5 +101,22 @@ func (reconciler *Reconciler) reconcileDebounced(ctx context.Context) error {
 		return err
 	}
+
+	if !reconciler.healthChecker.IsInitialized() {
```
Why are we doing all these checks if the healthChecker is not yet initialized?

Wouldn't we expect them to fail anyway?

p-strusiewiczsurmacki-mobica replied:
I think the method name was not descriptive enough. Changed it to `IsNetworkingHealthy()`.

The whole point of this check is, because reconciler runs in a loop, to run the healthcheck only if networking was not already checked and determined to be healthy, and set `isNetworkingHealhy` (formerly `isInitialized`) variable accordingly. After it is set, node taint should be removed and healthcheck should not be performed again.

watzkuh commented:
LGTM besides those very minor comments

Cellebyte reviewed (commented):
LGTM 

added hint that it closes the Issue raised regarding node discovery by hostname.
